ReSharper.FSharp/src/FSharp.Psi.Features/src/CodeCompletion/Rules/NamedUnionCaseFieldsPatRule.fs

```diff
@@ -0,0 +1,99 @@
+
+    // Find the parameterOwner `A` in `A()` or `A(a = a; )`
+    // Filter out the already used fields
+    let getFieldsFromParametersOwnerPat (parametersOwnerPat: IParametersOwnerPat) (filterFields: Set<string>) =
+        let referenceName = parametersOwnerPat.ReferenceName
+        if isNull referenceName then None else
+
+        // We need to figure out if `A` actually is a UnionCase.
+        let fcsUnionCase = referenceName.Reference.GetFcsSymbol().As<FSharpUnionCase>()
+        if isNull fcsUnionCase then None else
+
+        let fieldNames =
+            fcsUnionCase.Fields
+            |> Seq.choose (fun field -> if field.IsNameGenerated then None else Some field.Name)
+            |> Seq.toArray
+
+        if Set.isEmpty filterFields then Some fieldNames else
+        fieldNames
+        |> Array.except filterFields
+        |> Some
+
+    // The current scope is to have A({caret}) captured.
+    // A fake identifier will be inserted in the reparseContext.
+    let getFieldsFromReference (context: FSharpCodeCompletionContext) =
+        let parametersOwnerPat = getParametersOwnerPatFromReference context.ReparsedContext.Reference
+        if isNull parametersOwnerPat then None else
+
+        let filteredItems =
+            let namedUnionCaseFieldsPat = parametersOwnerPat.Parameters.SingleItem.As<INamedUnionCaseFieldsPat>()
+            if isNull namedUnionCaseFieldsPat then Set.empty else
+
+            namedUnionCaseFieldsPat.FieldPatterns
+            |> Seq.map (fun fieldPat -> fieldPat.ShortName)
+            |> Seq.filter (fun name -> name <> SharedImplUtil.MISSING_DECLARATION_NAME)
+            |> Set.ofSeq
+
+        getFieldsFromParametersOwnerPat parametersOwnerPat filteredItems
+
+    override this.IsAvailable(context) =
+        let fieldNames = getFieldsFromReference context
+        Option.isSome fieldNames
+
+    override this.TransformItems(context, collector) =
+        let parametersOwnerPat = getParametersOwnerPatFromReference context.ReparsedContext.Reference
+        let namedUnionCaseFieldsPat = if isNull parametersOwnerPat then null else parametersOwnerPat.Parameters.SingleItem.As<INamedUnionCaseFieldsPat>()
+        if isNotNull namedUnionCaseFieldsPat then
+            // In this scenario we are already in a named union case field pattern.
+            // Thus there is no need to show any other completions.
+            collector.RemoveWhere(fun item -> true)
+
+        let fieldNames = getFieldsFromReference context
+        match fieldNames with
+        | None -> ()
+        | Some fieldNames ->
+
+        for fieldName in fieldNames do
+            let info = TextualInfo(fieldName, fieldName, Ranges = context.Ranges)
+            let item =
+                LookupItemFactory
+                    .CreateLookupItem(info)
+                    .WithPresentation(fun _ ->
+                        TextPresentation(info, fieldName, true, PsiSymbolsThemedIcons.Field.Id))
```

Comment:
@nojaf The type is passed here.

Reply:
Thanks, I used the `FCS` API to get the type name (`field.FieldType.Format`).
I hope that is ok? Let me know if there is a better ReSharper API for this.

Reply:
Thanks!

Yes, that's correct and is what we in similar places. The reason is we want to have F# view of the types (e.g. `int option`, not CLR/R# one (`FSharpOption<int>`) here.